File: api/src/main/java/com/bindglam/mint/manager/Managerial.java
package com.bindglam.mint.manager;

import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

/**
 * Manager interface
 *
 * @author bindglam
 */
public interface Managerial {
    @ApiStatus.Internal
    default void preload(@NotNull Context context) {
    }

    @ApiStatus.Internal
    default void start(@NotNull Context context) {
    }

    @ApiStatus.Internal
    default void end(@NotNull Context context) {
    }

    @ApiStatus.Internal
    default @NotNull Priority priority() {
        return Priority.empty();
    }

    record Priority(int start, int end) {
        private static final Priority EMPTY = new Priority(0, 0);

        public static @NotNull Priority empty() {
            return EMPTY;
        }

        public static @NotNull Priority of(int start, int end) {
            return new Priority(start, end);
        }
    }
}

File: core/src/main/kotlin/io/github/bindglam/mint/manager/Managerial.kt
package io.github.bindglam.mint.manager

interface Managerial {
    fun preload(context: Context) {
    }

    fun start(context: Context) {
    }

    fun end(context: Context) {
    }

    fun priority(): Priority = Priority.empty()

    data class Priority(val start: Int, val end: Int) {
        companion object {
            private val EMPTY = Priority(0, 0)
            fun empty() = EMPTY

            fun of(start: Int, end: Int) = Priority(start, end)
        }
    }
}